### ISSUE TITLE
[Non-glslang] Various SPIR-V consumers crash when run under a SPIR-V fuzzer (which tools?)

### ISSUE THREAD

**Issue by astarasikov** [KhronosGroup/glslang#196]
Hello.
I decided to use a fuzzer to generate SPIR-V binaries that could crash the tools which work with them, and found several cases which either trigger assertions or very long execution times in "spirv-remap". I have also found out that a certain vendor driver also crashes when loading those binaries.

I wrote a blog post summarizing my findings here - http://allsoftwaresucks.blogspot.ru/2016/03/fuzzing-vulkans-how-do-they-work.html .

Would you like to have a look at the binaries (I have uploaded them)?

Do you think it would be useful to create a script to run fuzzing with both SPIR-V binaries and GLSL shaders to test the quality of the parser and other parts of the validator, should I do more testing?

I think it would be good to take a look at the cases which cause long compilation time on simple shaders, and also it would be nice to convert the assertions into graceful shutdowns.

**Comment by xorgy** [comment]
@astarasikov By the way, it may be courteous to at least clean up and minimize the test cases; else understanding them is pretty useless. Take a look at `afl-tmin`'s help and feel free to ask some questions.

**Comment by johnkslang** [comment]
Agreed with @xorgy that this class of tests is for a SPIR-V validator, or tools that are supported for accepting invalid SPIR-V.

I'd like to close this issue, as glslang is not such a tool.  Will leave open for a bit to identify what/where those tools are.

**Comment by xorgy** [comment]
WIth that said, nothing should prevent people from implementing failure conditions properly in these tools, which should make using them less painful. I would hate for the tool to always simply crash on bad data, even if i really wish the data were valid.

**Comment by johnkslang** [comment]
@xorgy agreed it's always nicer to have higher quality if a validator is bypassed. The point is that SPIR-V ecosystem is generally expecting a validator to be available and inserted into tool chains such that a large set of checking is done once there, rather than repeated N times in N tools.

**Comment by dneto0** [comment]
Hi @astarasikov
Please try fuzzing the spirv-dis and spirv-val tools from https://github.com/KhronosGroup/SPIRV-Tools

spirv-dis is a SPIR-V disassembler.  Its key component is a binary parser that is meant to be used in  other SPIR-V consumers.
spirv-val is a SPIR-V validator.  It does not yet implement all checks, but it should certainly not crash.

If you do find bugs in them, please file issues against KhronosGroup/SPIRV-Tools.

**Comment by astarasikov** [comment]
Hello @dneto0 
I have tried fuzzing both **spirv-dis**, **spirv-val** and **spirv-as**. I ran it for a longer period of time (2hours), and I tried using various shaders generated by **glslang** as the reference files for the fuzzer. It did not crash or freeze a single time.

I also tried feeding **spirv-as** with a large file (5Gb+) generated by taking a sample spirv-dis output and repeating one FunctionCall line multiple times (used the "cat contents >> contents2; cat contents2 >> contents" to quickly grow the file exponentially). It crashed with an OOM (my VM had 3Gb RAM which should be a reasonable configuration for a mobile device). I then repeated a similar test for **spirv-dis** and it also crashed (unsurprisingly since they have the same pattern for processing the input data).

I think the simplest way to deal with it would be introducing the limit on the input source code size. An option would be to catch the corresponding error, but then one may come up with a scenario where this would cause DoS by exhausting system memory.

```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted

```
I also liked how you have tests for most functions and carefully parse the input.
